src/celutil/hashmap.h

```diff
@@ -0,0 +1,225 @@
+
+#pragma once
+
+#include <cstddef>
+#include <algorithm>
+
+template<typename K, typename V>
+class HashMap
+{
+protected:
+    K *m_keys;
+    V *m_vals;
+    size_t m_size { 0 };
+    size_t m_used { 0 };
+    size_t m_shift { 0 };
+    size_t m_thres { 0 };
+    float m_loadF { 0.8 };
+    size_t m_maxSize { 1024 };
+    size_t m_minSize { 16 };
+    V m_zero;
+    bool m_hasZero { false };
+public:
+    static V invalidValue();
+    size_t size() const { return m_size + 1; }
+    size_t used() const { return m_hasZero ? m_used + 1 : m_used; }
+    const K *keyData() const { return m_keys; }
+    const V *valData() const { return m_vals; }
+    size_t hashedKey(K k) const
+    {
+        return (k ^ k >> (sizeof(k) * 4) * 0x9E3779B97F4A7C15L >> m_shift);
```

Comment:
where this magic came from?

Reply:
Adapted from https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/utils/LongMap.java
When in single function, gave wrong results.
Anyway, I'm again out of developing. Not sure, if I'll return again. You can forget this PR, if you have better things to do.

Reply:
that's sad